### PR TITLE
[SPARK-20665][SQL]"Bround"  and "Round" function return NULL

### PR DESCRIPTION
**Pull request by 10110346**
## What changes were proposed in this pull request?
   spark-sql>select bround(12.3, 2);
   spark-sql>NULL
For this case,  the expected result is 12.3, but it is null.
So ,when the second parameter is bigger than "decimal.scala", the result is not we expected.
"round" function  has the same problem. This PR can solve the problem for both of them.

## How was this patch tested?
unit test cases in MathExpressionsSuite and MathFunctionsSuite

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @10110346 .
What about `round`?

**Comment by 10110346** [comment]
"round" has the same problem.@ dongjoon-hyun . Actually, this PR can solve the problem for both of them

**Comment by dongjoon-hyun** [comment]
Then, let's change the title and description of this PR and the JIRA.
Also, could you add more test cases for `round`, too?

**Comment by 10110346** [comment]
ok, i will do it, thanks  @dongjoon-hyun

**Comment by 10110346** [comment]
Please reveiw  it，thanks   @dongjoon-hyun  @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala [hunk 1]
```diff
@@ -546,15 +546,14 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     val bdResults: Seq[BigDecimal] = Seq(BigDecimal(3.0), BigDecimal(3.1), BigDecimal(3.14),
       BigDecimal(3.142), BigDecimal(3.1416), BigDecimal(3.14159),
       BigDecimal(3.141593), BigDecimal(3.1415927))
-    // round_scale > current_scale would result in precision increase
-    // and not allowed by o.a.s.s.types.Decimal.changePrecision, therefore null
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala]: seems we are changing the behavior, can you check with other database and see if it's expected to return the original value for this case?

**Comment by 10110346** [comment]
@cloud-fan 
I have tested in mysql:
mysql> select round(12.3, 2);
+----------------+
| round(12.3, 2)    |
+----------------+
|          12.30        |
+----------------+

Also，I  tested in spark1.4, as follows:
> select round(12.3,2)
+-------+
|  _c0      |
+-------+
| 12.3      |
+-------+

**Comment by cloud-fan** [comment]
LGTM, cc @gatorsmile to double check

**Comment by gatorsmile** [comment]
ok to test

**Comment by gatorsmile** [comment]
Yeah. This is a right fix. I checked Hive and it behaves the same. 

```
hive> SELECT round(3.1415925, 7), round(3.1415925, 8), round(3.1415925, 9), round(3.1415925, 10), round(3.1415925, 100), round(3.1415925, 6), round(null, 8);
OK
3.1415925	3.1415925	3.1415925	3.1415925	3.1415925	3.141593	NULL
```

Before the fix, we return a wrong result. 
```
spark-sql> SELECT round(3.1415925, 7), round(3.1415925, 8), round(3.1415925, 9), round(3.1415925, 10), round(3.1415925, 100), round(3.1415925, 6), round(null, 8);
3.1415925	NULL	NULL	NULL	NULL	3.141593	NULL
```

File: sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala [hunk 2]
```diff
@@ -231,6 +231,19 @@ class MathFunctionsSuite extends QueryTest with SharedSQLContext {
...
+    checkAnswer(
+      sql(s"SELECT round($bdPi, 7), round($bdPi, 8), round($bdPi, 9), round($bdPi, 10), " +
+        s"round($bdPi, 100), round($bdPi, 6), round(null, 8)"),
+      Seq(Row(bdPi, bdPi, bdPi, bdPi, bdPi, BigDecimal("3.141593"), null))
+    )
+
+    checkAnswer(
+      sql(s"SELECT bround($bdPi, 7), bround($bdPi, 8), bround($bdPi, 9), bround($bdPi, 10), " +
+        s"bround($bdPi, 100), bround($bdPi, 6), bround(null, 8)"),
+      Seq(Row(bdPi, bdPi, bdPi, bdPi, bdPi, BigDecimal("3.141592"), null))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala]: How about moving the test case to `SQLQueryTestSuite` and create a new file `mathFunctions.sql`. You can refer to `s.c.s.t.resources/sql-tests/inputs/cast.sql`. To generate the result file, you can run the following command:
```
SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/test-only *SQLQueryTestSuite"
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala]: we can do it in a follow-up

**Comment by cloud-fan** [comment]
hi @10110346 , can you set the `affected versions` more accurate in the JIRA ticket? Does Spark 2.0 has the same issue? We rely on that information to decide which branches we want to backport this patch. Thanks!

**Comment by 10110346** [comment]
@cloud-fan  ok, I will do it

**Comment by 10110346** [comment]
@cloud-fan  Spark 2.0 and  Spark 2.1 have the same issue. I have updated the affected versions in the JIRA. Thanks!

**Comment by cloud-fan** [comment]
thanks, merging to master/2.2/2.1/2.0!

**Pull request closed without merging**